### ISSUE TITLE
Empty trees in a model with early stopping?

### ISSUE THREAD

**Issue by GuidoBartoli** [dmlc/xgboost#6448]
Hi, I trained a _XGBClassifier_ model on a 32-feature dataset using an early stop of 10 epochs on validation loss while giving 60 as the maximum number of trees. The fitting stopped at round 19, so the `Booster` object inside the saved model accordingly contains `best_ntree_limit = 19`.

When using the `plot_tree()` function, I noticed that the model has "normal" trees until round 24 (see below):
- Tree 18:
![18](https://user-images.githubusercontent.com/16103676/100631787-5f935200-332c-11eb-9d8e-280a9eeffe63.png)
- Tree 19:
![19](https://user-images.githubusercontent.com/16103676/100631791-602be880-332c-11eb-943e-631da37d8670.png)
- Tree 20:
![20](https://user-images.githubusercontent.com/16103676/100631792-602be880-332c-11eb-8eb2-92f85966dfb0.png)
- Tree 21:
![21](https://user-images.githubusercontent.com/16103676/100631794-60c47f00-332c-11eb-8d88-f995e5b71fa4.png)
- Tree 22:
![22](https://user-images.githubusercontent.com/16103676/100631797-60c47f00-332c-11eb-8b65-43d49a2c2562.png)
- Tree 23:
![23](https://user-images.githubusercontent.com/16103676/100631800-615d1580-332c-11eb-8e63-873937dbdbb9.png)
- Tree 24:
![24](https://user-images.githubusercontent.com/16103676/100631802-61f5ac00-332c-11eb-87f4-1eee75834f62.png)
But tree 25 is empty, i.e. it return directly 0:
![25](https://user-images.githubusercontent.com/16103676/100631804-61f5ac00-332c-11eb-9154-d508f4451850.png)
Subsequent tree are back to normal:
- Tree 26:
![26](https://user-images.githubusercontent.com/16103676/100631805-628e4280-332c-11eb-8f5c-fa4ee1c468a9.png)
- Tree 27:
![27](https://user-images.githubusercontent.com/16103676/100631807-628e4280-332c-11eb-99b8-f2da90c11c93.png)
- Tree 28:
![28](https://user-images.githubusercontent.com/16103676/100631808-6326d900-332c-11eb-8b10-7971475ecd63.png)
Tree 29 and following, instead, do not exist (that should be because early stopping occurred 10 epochs after tree 19).

I know that during prediction the estimation is automatically limited to 19 rounds according to the embedded `best_ntree_limit`, but out of curiosity I was wondering if it is normal to have empty trees inside a trained model, even with early stopping.

Thanks,
Best regards

**Comment by trivialfis** [comment]
If you are running 1.3rc, the early stopping callback has a `save_best` option to eliminate all the extra trees.  Also, the multi-class implementation is using 1 tree per class and aggregate the result with sorfmax as a 1 vs rest approach.  So `ntree_limit` is not a great way to slice the model.  In 1.3rc, there's a slice method, eg:

```
cls = XGBClassifier()
cls.fit(...)  # Put EarlyStopping callback here optionally
booster = cls.get_booster()
trees_in_first_iteration = booster[0: 1]
```

Then you can see each tree in each iteration.  An empty tree doesn't mean that boosting round is not doing anything, it just means during that boosting round, there's no gain in a specific class.

**Comment by trivialfis** [comment]
https://xgboost.readthedocs.io/en/latest/python/model.html

**Comment by GuidoBartoli** [comment]
Ok, thank you very much for the explanation! :)
Waiting for the official 1.3 release...
